src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -199,19 +187,42 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
 
                     // Check is a reference of any sort (i.e. read/write/nameof) to the local.
                     if (identifierName.Identifier.ValueText == localSymbol.Name)
-                    {
                         return;
-                    }
                 }
             }
 
             if (!Analyzer.CanSafelyConvertToPatternMatching(
-                semanticModel, localSymbol, comparison, operand,
-                localStatement, enclosingBlock, cancellationToken))
+                    semanticModel, localSymbol, comparison, operand,
+                    localStatement, enclosingBlock, cancellationToken))
             {
                 return;
             }
 
+            var comparisonEnclosingBlock = comparison.AncestorsAndSelf().FirstOrDefault(n => n is BlockSyntax);
+            if (comparisonEnclosingBlock is null)
+                return;
+
+            if (comparisonEnclosingBlock != enclosingBlock)
+            {
+                // ok, the local variable is defined in a different block than the block that the `x != null` is in. If
+                // we then update the `x != null` to `o is X x` we may break scoping if the variable is referenced
+                // before/after that scope.
+                foreach (var descendentNode in enclosingBlock.DescendantNodes())
+                {
+                    var descendentNodeSpanStart = descendentNode.SpanStart;
+                    if (descendentNodeSpanStart <= localStatementStart)
+                        continue;
+
+                    if (descendentNode is IdentifierNameSyntax identifierName && !
+                        descendentNode.Span.IntersectsWith(comparisonEnclosingBlock.Span) &&
```

Comment:
why is the intersectsWith check necessary?

Reply:
because the varaible *is* allowed under the scope of hte inner block.  we want to see if it's used *outside* of that block.